src/auth0-session/client/edge-client.ts

```diff
@@ -0,0 +1,241 @@
+          )
+        );
+      }
+      return {
+        signal: AbortSignal.timeout(this.config.httpTimeout),
+        headers
+      };
+    };
+  }
+
+  private async getClient(): Promise<[oauth.AuthorizationServer, oauth.Client]> {
+    if (this.as) {
+      return [this.as, this.client as oauth.Client];
+    }
+
+    const issuer = new URL(this.config.issuerBaseURL);
+    try {
+      this.as = await oauth
+        .discoveryRequest(issuer, this.httpOptions())
+        .then((response) => oauth.processDiscoveryResponse(issuer, response));
+    } catch (e) {
+      throw new DiscoveryError(e, this.config.issuerBaseURL);
+    }
+
+    this.client = {
+      client_id: this.config.clientID,
+      ...(!this.config.clientAssertionSigningKey && { client_secret: this.config.clientSecret }),
+      token_endpoint_auth_method: this.config.clientAuthMethod,
+      id_token_signed_response_alg: this.config.idTokenSigningAlg,
+      clockTolerance: this.config.clockTolerance
+    };
+
+    return [this.as, this.client];
+  }
+
+  async authorizationUrl(parameters: Record<string, unknown>): Promise<string> {
+    const [as] = await this.getClient();
+    const authorizationUrl = new URL(as.authorization_endpoint as string);
+    authorizationUrl.searchParams.set('client_id', this.config.clientID);
+    Object.entries(parameters).forEach(([key, value]) => {
+      if (value === null || value === undefined) {
+        return;
+      }
+      authorizationUrl.searchParams.set(key, String(value));
+    });
+    return authorizationUrl.toString();
+  }
+
+  async callbackParams(req: Auth0Request, expectedState: string) {
+    const [as, client] = await this.getClient();
+    const url =
+      req.getMethod().toUpperCase() === 'GET' ? new URL(req.getUrl()) : new URLSearchParams(await req.getBody());
+    let result: ReturnType<typeof oauth.validateAuthResponse>;
+    try {
+      result = oauth.validateAuthResponse(as, client, url, expectedState);
+    } catch (e) {
+      throw new ApplicationError(e);
+    }
+    if (oauth.isOAuth2Error(result)) {
+      throw new IdentityProviderError({
+        message: result.error_description || result.error,
+        error: result.error,
+        error_description: result.error_description
+      });
+    }
+    return result;
+  }
+
+  async callback(
+    redirectUri: string,
+    parameters: URLSearchParams,
+    checks: OpenIDCallbackChecks,
+    extras: CallbackExtras
+  ): Promise<TokenEndpointResponse> {
+    const [as, client] = await this.getClient();
+
+    const { clientAssertionSigningKey, clientAssertionSigningAlg } = this.config;
+
+    let clientPrivateKey = clientAssertionSigningKey as CryptoKey | undefined;
+    /* c8 ignore next 3 */
+    if (clientPrivateKey && !(clientPrivateKey instanceof CryptoKey)) {
+      clientPrivateKey = await jose.importPKCS8<CryptoKey>(clientPrivateKey, clientAssertionSigningAlg || 'RS256');
+    }
+    const response = await oauth.authorizationCodeGrantRequest(
```

Comment:
Can this call throw an exception? If so, should it be caught, wrapped, and rethrown?

Reply:
This gets caught here https://github.com/auth0/nextjs-auth0/blob/edge-auth-handlers/src/auth0-session/handlers/callback.ts#L61-L79

Reply:
In the node handler we're wrapping the errors in our own, more specific error types:

https://github.com/auth0/nextjs-auth0/blob/141ec7bff57684747e99baddf02840706adda33b/src/auth0-session/client/node-client.ts#L188

Reply:
`authorizationCodeGrantRequest` doesn't throw an `IdentityProvider` error (one with an `error` / `error_description` that comes from the IDP) it just returns a fetch response (or throws config errors like, "you forgot the `response_type`")

`processAuthorizationCodeOpenIDResponse` returns an IDP error (not throws) which is thrown on L158